### ISSUE TITLE
Keras/Tensorflow train_on_batch memory leak

### ISSUE THREAD

**Issue by Emile0205** [tensorflow/tensorflow#44728]
Please do help

train_on_batch has a memory leak in my and many other people's code. It was thought to only be on predict and fit, which was solved with a del and gc.collect(), but train_on_batch cannot be solved by this. Please someone help!

**Comment by Saduf2019** [comment]
@Emile0205 
Please share a simple stand alone code to replicate the issue reported.

**Comment by bguetarni** [comment]
Hi @Emile0205,
just to tell you that I have found a solution since last time for this, but it works for me, maybe not for everyone.
I just wrote my train loop manually as indicated in [Writing a training loop from scratch](https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch#low-level_handling_of_losses_tracked_by_the_model).
Be carefull about how you send your data to your model if you use a GPU, remember that the way you index a **numpy** array will impact the amount of memory your GPU allocates (prefer using **copy**)!
Hope it will help you !

**Comment by Emile0205** [comment]
Good day @bguetarni !!!! That link is amazing, thank you! I really enjoy the idea of taking more control of the loss functions and gradients than using Model and compile (ugh!).... I am, however, struggling with altering the GAN example code in the shared link to allow for feature matching  - i.e. how can I access a sequential model's final output and some intermediate layer output as well. Any help from anyone will be greatly appreciated!!

**Comment by galchinsky** [comment]
I constantly meet the bug and it's hard to catch because the leak is very slow and sometimes it disappears after touching the model code, sometimes it is slow enough to finish the training, sometimes it dies after few hours eating 64 Gb of RAM. Turning eager mode off helped to make it slower. There was also an option to try `fit(...epochs=1...)` but I don't remember if this helps.
But I confirm that it is inside `train_on_batch` and still persists on tf2.3.1 from pip. I don't remember it in tf1.

As @bguetarni says, the only 100% workaround was to use a custom training step instead of train_on_batch. Note that I had to implement LossScaleOptimizer for mixed precision training after reading the original code of fit/train_on_batch/make_train_function and this code is better in nuances than the documentation.
Most Internet tutorials like https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch doesn't have this, so you may meet a weird loss function behavior when turn mixed precision on.

```
from tensorflow.python.keras.mixed_precision.experimental import loss_scale_optimizer as lso
@tf.function                                                                                
def train_step(inputs, targets):                                                                                                          
    with tf.GradientTape() as tape:                                                         
        preds = model(inputs, training=True)                                      
        loss_value = model.compiled_loss(targets, preds)                          
        scaled_loss_value = loss_value                                                      
        if isinstance(model.optimizer, lso.LossScaleOptimizer):                   
            scaled_loss_value = model.optimizer.get_scaled_loss(scaled_loss_value)
    grad = tape.gradient(scaled_loss_value, model.trainable_variables)            
    model.optimizer.apply_gradients(zip(grad, model.trainable_variables))     
    return loss_value                                  
```

This is basically a drop-in replacement for train_on_batch: `loss_gen = train_step(xbatch, ybatch)` so I'm quite happy now with this.
